app/Http/Controllers/ListingController.php

```diff
@@ -0,0 +1,137 @@
+        $tags = Tag::all();
+        $categories = Category::all();
+
+        return view('listings.create', compact('tags', 'categories'));
+    }
+
+    /**
+     * Store a newly created listing in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function store(Request $request)
+    {
+        $listing = Listing::createListing($request);
+
+        // Attach the selected category to the listing
+        $listing->categories()->attach($request->input('category'));
+
+        return redirect()->route('listings.index')->with('success', 'Listing created successfully!');
+    }
+
+    /**
+     * Show the form for editing the specified listing.
+     *
+     * @param  \App\Models\Listing  $listing
+     * @return \Illuminate\Http\Response
+     */
+    public function edit(Listing $listing)
+    {
+        // Check if the authenticated user is the creator of the listing
+        if (Gate::denies('update-listing', $listing)) {
+            abort(403, 'Unauthorized');
+        }
+
+        // Retrieve the list of available tags and categories
+        $tags = Tag::all();
+        $categories = Category::all();
+
+        return view('listings.edit', compact('listing', 'tags', 'categories'));
+    }
+
+    /**
+     * Update the specified listing in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \App\Models\Listing  $listing
+     * @return \Illuminate\Http\Response
+     */
+    public function update(Request $request, Listing $listing)
+    {
+        // Check if the authenticated user is the creator of the listing
+        if (Gate::denies('update-listing', $listing)) {
+            abort(403, 'Unauthorized');
+        }
+
+        $request->validate([
+            'title' => 'required|max:255',
+            'description' => 'required',
+            'tags' => 'array', // Ensure 'tags' is an array in the request
+        ]);
+
+        $listing->update([
+            'title' => $request->input('title'),
+            'description' => $request->input('description'),
+        ]);
+
+        // Sync tags for the listing
+        $listing->tags()->sync($request->input('tags'));
+
+        // Update the selected category for the listing
+        $listing->categories()->sync($request->input('category'));
+
+        return redirect()->route('listings.index')->with('success', 'Listing updated successfully!');
+    }
+
+    /**
+     * Remove the specified listing from storage.
+     *
+     * @param  \App\Models\Listing  $listing
+     * @return \Illuminate\Http\Response
+     */
+    public function destroy(Listing $listing)
+    {
+        // Check if the authenticated user is the creator of the listing
+        if (Gate::denies('delete-listing', $listing)) {
+            abort(403, 'Unauthorized');
+        }
+
+        // Update the 'deleted_by' column with the ID of the user who deleted the listing
+        $listing->update([
```

Comment:
to override the model softdelete

Listing.php
    /**
     * Trigger eloquent events to delete.
     *
     * @return void
     */
    public function softDelete()
    {
        $this->deleted_at = $this->freshTimestamp();
        $this->deleted_by= Auth::id;
        $this->timestamps = false;
        $this->save();
    }